### ISSUE TITLE
`schedule` cannot access local bindings

### ISSUE THREAD

**Issue by helins** [Convex-Dev/convex#354]
Given the current implementation, it is very easy reaching a situation where a scheduled transaction fails silently. Notably, the following example is counter-intuitive:

```clojure
(let [x 42] (expand '(schedule 0 (def foo x))))

;; (schedule* 0 (compile (quote (def foo x))))
;; Of course, `x` is not defined anymore ; fails silently
```

I don't think this is fixable given that nothing is supposed to be evaluated in advance. Maybe we should remove `schedule` and use only explicit templating with `schedule*`.

**Comment by mikera** [comment]
Interesting. Making the schedule contain an anonymous function would potentially solve this, with a slight overhead.

**Comment by helins** [comment]
Different from #353 but sounds like both should be coordinated.

I think I'd fancy templating more than a lambda:

- Don't have to store a closure (and pay for what you don't need)
- More in line with stuff like `deploy`

**Comment by mikera** [comment]
Yeah you can template / build whatever you like with `schedule*`. The question I guess is how to make the helper macro `schedule` most intuitive. 

If we want to preserve things like let bindings I think we have two main options:
1. Use a lambda. Easy since the support for locals is already there, but does add a bit of cost
2. Figure out some way to embed let bindings in `compile` so that they don't need a lookup lookup later.

Option 2 is probably optimal, just needs a bit of careful though on the semantics. Should compile *always* inline let-bound variables, for example?

**Comment by mikera** [comment]
Blocked on #413 (protocol upgrade mechanism). Any change to schedule semantics alters CVM transition behaviour and would change the genesis hash if applied naively — must ship as a version-gated upgrade. See [UPGRADE.md](../blob/develop/UPGRADE.md).

**Comment by mikera** [comment]
(Claude Comment) Status: confirmed still-present on `develop` — the `schedule` macro expands to `(schedule* ~timestamp (compile (quote ~code)))`, so surrounding local bindings compile as lookups that are unbound at execution time. Neither proposed fix (closure capture or local inlining) has been applied. This is now unblocked in principle: changing `schedule` semantics is a CVM behaviour change, and the network upgrade mechanism (#413) exists to carry it — this is a candidate for the v2 protocol migration. The existing `testScheduleExecution` only schedules a literal, so the fix should come with a local-binding regression test.
